### DISCUSSION TITLE
looking for suggestions for db modeling

### DISCUSSION THREAD

**Discussion by chaotic-justice** [prisma/prisma#5156] [Q&A] [answered]
I'm working on a backend w/ potentially 2 types of User: admin & applicant. Admins have permission to create a survey entity w/ many questions (responses could be string, date, or number). Applicants in turn can submit survey responses -> creating applications (for scholarships for instance) in the db. How should I go about designing the db? I see some use cases I can borrow from the grading app, the challenge is that whenever an admin creates a survey form, I'd have to hardcode the questions and the response types for graphql to digest its schema. Also, reviewers/admins have right to flag these applications as pass/fail.

**Comment by 2color** [comment, marked as answer]
Hey Chao,


There are two concerns to be addressed in your question:
- How to model the DB for a survey app.
- How to model the DB in a way that allows you to enforce permissions.

## Modeling an admin user

Since your data model is likely going to have a `User` model/table, you could add an `isAdmin` boolean field to mark admin users:
```prisma
model User {
  email   String  @unique
  id      Int     @id @default(autoincrement())
  name    String?
  isAdmin Boolean @default(false)
  // any other fields...
}
```

If you had more than two user types (admin and normal user), you could define an `UserType` enum and add that as a field on the `User` model.

## Modeling the survey app

The following data model creates a relation (1:n) between `User` and `Survey`. This means that any user can create multiple surveys (each with many questions and answers). 

```prisma
model User {
  email         String         @unique
  id            Int            @id @default(autoincrement())
  name          String?
  isAdmin       Boolean        @default(false)
  surveys       Survey[]
  surveyAnswers SurveyAnswer[]
}

model Survey {
  id        Int            @id @default(autoincrement())
  creator   User           @relation(fields: [creatorId], references: [id])
  creatorId Int
  questions Question[]
  answers   SurveyAnswer[]
}

model Question {
  id       Int    @id @default(autoincrement())
  survey   Survey @relation(fields: [surveyId], references: [id])
  surveyId Int
  question String
}

model SurveyAnswer {
  id         Int      @id @default(autoincrement())
  user       User     @relation(fields: [userId], references: [id])
  userId     Int
  question   Question @relation(fields: [questionId], references: [id])
  questionId Int      
  answer     String
}
```

Note that you can also add a relation between `SurveyAnswer` and `Survey`. In the schema above you that link between answers and surveys happens through the `Quesion` model.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
